openapiGenerator.go

```diff
@@ -389,15 +420,139 @@ func (g *openapiGenerator) generateMessageSchema(message *protomodel.MessageDesc
 	}
 	o := openapi3.NewObjectSchema()
 	o.Description = g.generateDescription(message)
+	msgRules := g.validationRules(message)
+	g.markerRegistry.MustApplyRulesToSchema(msgRules, o, markers.TargetType)
 
+	oneOfFields := make(map[int32][]string)
 	for _, field := range message.Fields {
+		repeated := field.IsRepeated()
+		fieldName := g.fieldName(field)
+		fieldDesc := g.generateDescription(field)
+		fieldRules := g.validationRules(field)
+
+		schemaType := g.markerRegistry.GetSchemaType(fieldRules)
+		if schemaType != "" {
+			tmp := getSoloSchemaForMarkerType(schemaType)
+			schema := getSchemaIfRepeated(&tmp, repeated)
+			schema.Description = fieldDesc
+			g.markerRegistry.MustApplyRulesToSchema(fieldRules, schema, markers.TargetType)
+			o.WithProperty(fieldName, schema)
+			continue
+		}
+
 		sr := g.fieldTypeRef(field)
-		o.WithProperty(g.fieldName(field), sr.Value)
+		g.markerRegistry.MustApplyRulesToSchema(fieldRules, sr.Value, markers.TargetType)
+		o.WithProperty(fieldName, sr.Value)
+
+		// If the field is a oneof, we need to add the oneof property to the schema
+		if field.OneofIndex != nil {
+			idx := *field.OneofIndex
+			oneOfFields[idx] = append(oneOfFields[idx], fieldName)
+		}
+	}
+
+	if g.protoOneof {
+		// Add protobuf oneof schema for this message
+		oneOfs := make([][]*openapi3.Schema, len(oneOfFields))
+		for idx := range oneOfFields {
+			// oneOfSchemas is a collection (not and required schemas) that should be assigned to the schemas's oneOf field
+			oneOfSchemas := newProtoOneOfSchema(oneOfFields[idx]...)
+			oneOfs[idx] = append(oneOfs[idx], oneOfSchemas...)
+		}
+
+		switch len(oneOfs) {
+		case 0:
+			// no oneof fields
+		case 1:
+			o.OneOf = getSchemaRefs(oneOfs[0]...)
+		default:
+			// Wrap collected OneOf refs with AllOf schema
+			for _, schemas := range oneOfs {
+				oneOfRef := openapi3.NewOneOfSchema(schemas...)
+				o.AllOf = append(o.AllOf, oneOfRef.NewRef())
+			}
+		}
 	}
 
 	return o
 }
 
+func getSoloSchemaForMarkerType(t markers.Type) openapi3.Schema {
+	switch t {
+	case markers.TypeObject:
+		return specialSoloTypes["google.protobuf.Struct"]
+	case markers.TypeValue:
+		return specialSoloTypes["google.protobuf.Value"]
+	}
+	return openapi3.Schema{}
```

Comment:
this `return` should never happen if i'm reading this right. should we panic?

Reply:
This return will be executed if we add a new Type enum in type.go, and that gets referenced in a proto file. My thought was to just ignore it, but we can panic and return